watertap/property_models/ion_DSPMDE_prop_pack.py

```diff
@@ -0,0 +1,838 @@
+                raise RuntimeError("While using the calculate_state method on {sb_name}, the degrees "
+                                   "of freedom were {dof}, but 0 is required. Check var_args and ensure "
+                                   "the correct fixed variables are provided."
+                                   "".format(sb_name=sb.name, dof=degrees_of_freedom(sb)))
+
+        # Solve
+        with idaeslog.solver_log(solve_log, idaeslog.DEBUG) as slc:
+            results = solve_indexed_blocks(opt, [self], tee=slc.tee)
+            solve_log.info_high("Calculate state: {}.".format(idaeslog.condition(results)))
+
+        if not check_optimal_termination(results):
+            _log.warning("While using the calculate_state method on {sb_name}, the solver failed "
+                         "to converge to an optimal solution. This suggests that the user provided "
+                         "infeasible inputs, or that the model is poorly scaled, poorly initialized, "
+                         "or degenerate.")
+
+        # unfix all variables fixed with var_args
+        for (k, v_name, ind), previously_fixed in flags.items():
+            if not previously_fixed:
+                var = getattr(self[k], v_name)
+                var[ind].unfix()
+
+        # fix state variables if hold_state
+        if hold_state:
+            fix_state_vars(self)
+
+        return results
+
+@declare_process_block_class("DSPMDEStateBlock",
+                             block_class=_DSPMDEStateBlock)
+class DSPMDEStateBlockData(StateBlockData):
+    def build(self):
+        """Callable method for Block construction."""
+        super(DSPMDEStateBlockData, self).build()
+
+        self.scaling_factor = Suffix(direction=Suffix.EXPORT)
+
+        # Add state variables
+        self.flow_mol_phase_comp = Var(
+            self.params.phase_list,
+            self.params.component_list,
+            initialize=100, #todo: revisit
+            bounds=(1e-8, None),
+            domain=NonNegativeReals,
+            units=pyunits.mol/pyunits.s,
+            doc='Mole flow rate')
+
+        self.temperature = Var(
+            initialize=298.15,
+            bounds=(273.15, 373.15),
+            domain=NonNegativeReals,
+            units=pyunits.degK,
+            doc='State temperature')
+
+        self.pressure = Var(
+            initialize=101325,
+            bounds=(1e5, 5e7),
+            domain=NonNegativeReals,
+            units=pyunits.Pa,
+            doc='State pressure')
+
+    # -----------------------------------------------------------------------------
+    # Property Methods
+    def _mass_frac_phase_comp(self):
+        self.mass_frac_phase_comp = Var(
+            self.params.phase_list,
+            self.params.component_list,
+            initialize=lambda b,p,j : 0.4037 if j == "H2O" else 0.0033, #todo: revisit
+            bounds=(1e-6, None),  # upper bound set to None because of stability benefits
```

Comment:
In IDAES, we have generally started making the upper bound 1.001 or something similar (i.e. have an upper bound, but give it a little slack).

Reply:
OK, I'll add upper bounds instead of None.